Hi mentors!

nanemo opened the pull request:
Web application is created. Now i am working on Session, Servlet Testing and CSS. Application works on 14 JDK and on Tomcat 9.0.48 version. Program works through JDBC. You can see all abst classes about DAO in service package.

I will check all my code for neatness at the end via SonarLint )).

Commit:
Hi mentors!
Web application is created. Now i am working on Session, Servlet Testing and CSS.
Application works on 14 JDK and on Tomcat 9.0.48 version.
Program works through JDBC. You can see all abst classes about DAO in service package.

I will check all my code for neatness at the end via SonarLint )).

nanemo commented:
Yuriy and Andrey, my web app works via JDBC, but i have a problem. I don't know how to push MySql tables with project code to GitHub that you could run project together both of them.

nanemo commented:
![image](https://user-images.githubusercontent.com/71605950/203128354-e97ddd0e-9d59-4d91-8b5f-e233fd5f3537.png)

Commit:
adding pom, docker, maven-wrapper files

Anderwerty commented on pom.xml:
```diff
@@ -0,0 +1,84 @@
+        <dependency>
+            <groupId>javax.servlet</groupId>
+            <artifactId>javax.servlet-api</artifactId>
+            <version>4.0.1</version>
```
please extract this version into maven properties
the same about other dependencies

nanemo replied:
problem solved

Anderwerty commented on pom.xml:
```diff
@@ -0,0 +1,84 @@
+            <artifactId>mysql-connector-java</artifactId>
+            <version>8.0.28</version>
+        </dependency>
+        <dependency>
```
do we need it as dependency?

nanemo replied:
deleted

Anderwerty commented on src/main/java/ru/javarush/quest/controller/LogicServlet.java:
```diff
@@ -0,0 +1,78 @@
+
+    @Override
+    protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws ServletException, IOException {
+        if (count > 3) count = 0;
```
`3` is magic number

nanemo replied:
problem solved. for number 3 i added context.getQuestionSize() method. It is game level size.

Anderwerty commented on src/main/java/ru/javarush/quest/controller/LogicServlet.java:
```diff
@@ -0,0 +1,78 @@
+        return context.getAnswer(count);
+    }
+
+    private Question getQuestion() {
```
logic should be in service

nanemo replied:
problem solved

nanemo replied:
I took both of them to Contex class

Anderwerty commented on src/main/java/ru/javarush/quest/entity/Account.java:
```diff
@@ -0,0 +1,64 @@
+    private final String login;
+    private String password;
+
+    public Account(Long id, String name, String login, String password) {
```
We have 4 fields. Let's make all fields final and use builder pattern

nanemo replied:
What is the different will be if i add builder pattern here? and why i have to add just in Account class no in Answer or Question classes?

Anderwerty commented on src/main/java/ru/javarush/quest/entity/Account.java:
```diff
@@ -0,0 +1,64 @@
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
```
https://google.github.io/styleguide/javaguide.html#s4.1.1-braces-always-used
Braces are used with if, else, for, do and while statements, even when the body is empty or contains only a single statement.

nanemo replied:
i deleted this equals method.

Anderwerty commented on src/main/java/ru/javarush/quest/entity/Account.java:
```diff
@@ -0,0 +1,64 @@
+    }
+
+    @Override
+    public int hashCode() {
```
let's use `Objects.hash` for simplification

nanemo replied:
i deleted this hash method.

Anderwerty commented on src/main/java/ru/javarush/quest/entity/Account.java:
```diff
@@ -0,0 +1,64 @@
+package ru.javarush.quest.entity;
+
+public class Account {
```
You have already added lombok. Why not use it for toString, equals and hashCode methods generation

nanemo replied:
I deleted lombok. I dont use already.

Anderwerty commented on src/main/java/ru/javarush/quest/impl/AnswerImpl.java:
```diff
@@ -0,0 +1,62 @@
+import java.sql.SQLException;
+import java.sql.Statement;
+
+public class AnswerImpl extends AbstractDao implements AnswerDao {
```
Maybe it is dao?

nanemo replied:
I replaced all names about Impl classes to Dao.

Anderwerty commented on src/main/java/ru/javarush/quest/impl/AnswerImpl.java:
```diff
@@ -0,0 +1,62 @@
+        try (Connection connection = connect();
+             Statement stmt = connection.createStatement()) {
+            stmt.execute("SELECT * from answers where id=" + count);
+            ResultSet resultSet = stmt.getResultSet();
```
it seems that https://docs.oracle.com/javase/7/docs/api/java/sql/ResultSet.html implements Autoclosable

nanemo replied:
ResultSet close method was added in finally block

Anderwerty commented on src/main/java/ru/javarush/quest/impl/AnswerImpl.java:
```diff
@@ -0,0 +1,62 @@
+                wrongAnswerDescription = getWrongAnswerDescription(rs);
+            }
+        } catch (Exception e) {
+            e.printStackTrace();
```
It is ok just log exception and return null?

nanemo replied:
What do you here. i didnt get it

Anderwerty commented on src/main/java/ru/javarush/quest/impl/LoginImpl.java:
```diff
@@ -0,0 +1,75 @@
+import java.sql.*;
+
+
+public class LoginImpl extends AbstractDao implements LoginDao {
```
the same

nanemo replied:
replaced

Anderwerty reviewed (commented):
It is also expected to write tests for your dao layer

Anderwerty commented on docker-compose.yml:
```diff
@@ -0,0 +1,30 @@
+version:
+services:
```
Sorry, but what functionality will create a database structure?

nanemo replied:
Please look at here again. About docker helped to me from our group Marina.

Commit:
I took my application on Docker, it works.
But i have uncompleted issues - Tests, CSS and Andrey's comments for my last committing.

Commit:
I took my application on Docker, it works.
But i have uncompleted issues - Tests, CSS and Andrey's comments for my last committing.

Commit:
I created CSS, Session and Test class for AnswerDao. Project is ready.

nanemo commented:
You can check my last commit in dev branch.
